File: Enhancement One New/AppointmentService.py
from datetime import datetime
import random
import Appointment


class AppointmentService:
    def __init__(self):
        self.appointment_list = []
        self.appointment_count = 0

    # getters
    def get_appointment_list(self):
        return self.appointment_list

    def get_appointment_count(self):
        return self.appointment_count

    # add appointments
    def add_appointment(self, date, description):
        today = datetime.now()
        if date is None or date < today:
            raise ValueError("Invalid Date")
        if description is None or len(description) > 50:
            raise ValueError("Invalid Description")

        new_id = self.generate_unique_id()
        new_appointment = Appointment(new_id, date, description)
        self.appointment_list.append(new_appointment)
        self.appointment_count += 1

    # delete appointment
    def delete_appointment(self, appointment_id):
        if not self.appointment_list:
            print("There are no appointments to delete")
            return

        if appointment_id is None or len(appointment_id) > 10:
            raise ValueError("Invalid ID")

        index = -1
        for idx, appointment in enumerate(self.appointment_list):
            if appointment.get_id() == appointment_id:
                index = idx
                break

        if index == -1:
            print("Appointment ID not found")
        else:
            del self.appointment_list[index]
            self.appointment_count -= 1
            print("Appointment removed")

    # generate a unique id for the appointment
    def generate_unique_id(self):
        rand = random.Random()
        new_id = rand.randint(0, 999999999)
        unique_id = str(new_id)

        id_list = [appointment.get_id() for appointment in self.appointment_list]
        while unique_id in id_list or len(unique_id) > 10:
            new_id = rand.randint(0, 999999999)
            unique_id = str(new_id)

        return unique_id

File: Enhancement One Original/projectpackage/AppointmentService.java
package projectpackage;

import java.util.Date;
import java.util.Random;
import java.util.Vector;

public class AppointmentService {
	
	private Vector<Appointment> appointmentList = new Vector<Appointment>();
	private int appointmentCount = 0;
	
	//getters
	public Vector<Appointment> GetAppointmentList() {
		return appointmentList;
	}
	public int GetAppointmentCount() {
		return appointmentCount;
	}
	
	//add appointments
	public void AddAppointment(Date date, String description) {
		Date today = new Date();
		if(date == null || date.before(today)) {
			throw new IllegalArgumentException("Invalid Date");
		}
		if (description == null || description.length() > 50) {
			throw new IllegalArgumentException("Invalid Description");
		}
		
		String newID = GenerateUniqueID();
		var newAppointment = new Appointment(newID, date, description);
		appointmentList.add(newAppointment);
		appointmentCount = GetAppointmentCount() + 1;
	}
	//delete appointment
	public void DeleteAppointment(String id) {
		if(appointmentList.isEmpty()) {
			System.out.println("There are no appointments to print");
		}
		if (id == null || id.length() > 10) {
			throw new IllegalArgumentException("Invalid ID");
		}
		//search for id
		int index = -1;
		for (Appointment a : appointmentList) {
			if (a.GetID() == id) {
				index = appointmentList.indexOf(a);
			}
		}
		//no appointment found
		if (index == -1) {
			System.out.println("Appointment ID no found");
			return;
		}
		else {
			//remove appointment
			appointmentList.remove(index);
			appointmentCount = GetAppointmentCount() - 1;
			System.out.println("Appointment removed");
		}
	}
	//make a unique id for the appointment
	public String GenerateUniqueID() {
		Random rand = new Random();
		int newID = rand.nextInt(1000000000);
		String uniqueID = Integer.toString(newID);
		
		Vector<String> idList = new Vector<String>();
		for (Appointment a : appointmentList) {
			idList.add(a.GetID());
		}
		while (idList.contains(uniqueID) || uniqueID.length() > 10) {
			newID = rand.nextInt(1000000000);
			uniqueID = Integer.toString(newID);
		}
		idList = null;
		return uniqueID;
	}

}
